src/main/java/io/jenkins/pluginhealth/scoring/service/PluginService.java

```diff
@@ -43,4 +46,16 @@ public Plugin saveOrUpdate(Plugin plugin) {
             .map(pluginRepository::save)
             .orElseGet(() -> pluginRepository.save(plugin));
     }
+
+    public List<Plugin> batchUpdate(List<Plugin> pluginList) {
+        for (int i = 0; i < pluginList.size(); i++) {
+            Optional<Plugin> pluginFromDatabase = pluginRepository.findByName(pluginList.get(i).getName());
+
+            if (pluginFromDatabase.isPresent()) {
+                pluginList.set(i, pluginFromDatabase.get().setScm(pluginList.get(i).getScm()).setReleaseTimestamp(pluginList.get(i).getReleaseTimestamp()));
+            }
+        }
+
+        return pluginRepository.saveAll(pluginList);
+    }
```

Comment:
Here, you are transforming the same list that you are iterating on. This cannot be correct and you should have concurrent access issues running this code.

However, you could leverage `Stream` API to find a plugin in DB, if it exists, then set the fields you want and finally collect that into a list that you send to the `Repository#saveAll` method.

Reply:
Thanks for pointing out the mistake related to concurrent access, I totally forgot about it.
I'm now using `Stream`  API in the new PR, please find it here: https://github.com/jenkins-infra/plugin-health-scoring/pull/26/files